Integrate combinable DFA capture resolution for our PCRE dialect regexes, part 1.

silentbicycle opened the pull request:
This adds an extra compilation step for PCRE regexes, generating programs for a group capture resolution abstract machine. The implementation builds on the approach described in ["Regular Expression Matching: the Virtual Machine Approach"](https://swtch.com/~rsc/regexp/regexp2.html), with adaptations to reduce the runtime memory footprint, handle some edge cases the same way PCRE does, explicitly detect some other PCRE edge cases and reject them as unsupported, and continue supporting group captures as the resulting DFAs are combined with other DFAs into a single DFA that matches them all in one pass

I have fuzzed this code quite a lot, both comparing group capture results to PCRE's and comparing libfsm's handling of individual regexes' DFAs with the DFA produced my combining them. Many of the test cases in `tests/capture/capture_test_case_list.c` came from fuzzing. My working branch diverged from `main` for a while during this work, in particular the metadata associated with end states, but after integrating upstream changes I fuzzed it further. I think I integrated things properly, but it's something to be mindful of during review. Some unrelated issues discovered during fuzzing have already been posted as separate PRs.

Worst-case memory usage is proportional to the size of the capvm program's length (known at compile time) and the input length -- we record a bit for each branch taken while evaluating the regex (e.g. either greedily jumping to repeat a subexpression or non-greedily advancing in the regex), so memory usage will slowly increase as inputs are evaluated. Common path prefixes are shared between threads, and the total thread count (and divergence) is bounded by the size of the opcode program. As a special case, a long path prefix of all zero bits is collapsed down to a counter; this usually happens because of an unanchored start loop, where each zero bit represents a path continuing to advance through the input without starting the match yet.

This PR targets an integration branch because some work is not yet complete:

- Some quick cleanup of CI issues -- the submodule's build config is more strict about variables that are set but unused (because their use is in logging code that compiles away).
- A few pathological cases having to do with repetition in combination with anchors are rejected as unsupported, such as `^(()($)|x)+$`. These are probably not worth the trouble to support. It also doesn't support the `\z` operator yet.
- Some of the combinations of CLI flags for `re` aren't enabled yet, such as capture resolution for multiple files, and the `re` docs need updating for some new flags.
- We should be able to calculate likely memory usage info at regex compile time, which could be used by the caller to stack-allocate and reuse a preallocated buffer rather than using dynamic allocation at runtime, but I haven't implemented that yet. The memory usage should small in practice, but it would be nice to completely eliminate runtime dynamic allocation. The path metadata table currently grows on demand, the other data structures are either fixed-size or the same length as the opcode count. Cases where it would hit the caller's memory limit for growing the path table are likely to hit PCRE's limits much sooner.
- Code generation does not output the capture resolution abstract machine's opcodes yet. When using capture handling, libfsm's caller will need to either link a small implementation of the abstract machine or have its code generation produce a standalone implementation, but because that brings some architectural trade-offs I wanted to wait until this round has been reviewed.

silentbicycle commented:
I've merged the current `main` into this and fuzzed it for another 20 minutes, on 8 CPU cores.

katef commented on fuzz/target.c:
```diff
@@ -23,17 +24,47 @@
+	MODE_REGEX,
+	MODE_REGEX_SINGLE_ONLY,
+	MODE_REGEX_MULTI_ONLY,
+	MODE_IDEMPOTENT_DET_MIN,
```
Can you add these new modes to ci.yml too please? where currently we have:
```
mode: [ m, p, d ]
```

silentbicycle replied:
Updated in 239927c3.

katef commented on fuzz/target.c:
```diff
@@ -143,21 +290,1069 @@ codegen(const struct fsm *fsm)
+
+static int
+compare_with_pcre(const char *pattern, struct fsm *fsm)
+{
```
fantastic

katef commented on fuzz/target.c:
```diff
@@ -369,12 +1649,87 @@ harness_fuzzer_target(const uint8_t *data, size_t size)
+			/* This triggers an "unreached" assertion in the parser.
+			 * It's already been reported (issue #386), but once the
+			 * fuzzer finds it, it will report it over and over.
+			 * Exit here so that the fuzzer considers it uninteresting. */
```
#386

silentbicycle committed:
bitmap.h: Add operations: copy, intersect, union, any, unset..

Most of these are used in fsm_detect_required_characters.

Also add #includes for standard headers being used.

silentbicycle committed:
Add fsm_detect_required_characters.

This inspects the DFA to determine which characters must appear in
any matching input.

katef committed:
Clarification, resolving a few TODOs, smaller CLI options etc.

katef committed:
Allocate and duplicate for \0-terminated patterns, rather than pointing at \n-terminated patterns in-situ. This is just so much less complicated.

katef committed:
Simplification, deal with -n earlier.

katef committed:
Give an example for AMBIG_ERROR.

I think at the time of writing, fsm_example() is broken, because I get no output here. Possibly related to #438

katef committed:
Move declined-file to an optarg, we need to keep argv[] for multiple input files.

katef committed:
Store patterns per set, rather than indexing into a global pattern array.

This sets the scene for multi-file pattern lists, later on. But also seems to keep scopes tighter and make things simpler as a byproduct.

katef committed:
Check end id ambiguity before printing, rather than during.

katef committed:
No need for the *nwritten parameter.

Now this is always equivalent to the count passed in, when the return status is 1. And the return status is always 1 when the count is enough. In all situations we know the count is enough.

katef committed:
Switch to reading and allocating line-by-line, rather than mmap().

This allows for eventually iterating over multiple input files.

katef committed:
C99 doesn't allow a different type qualifier here, only C2x.

katef committed:
Multi-file input.

When only one file is given, the endid is per pattern (that is, the line number) in the input file.

When multiple files are given, the endid is the argv[] index the file. That is, all patterns within each file share the same endid.

katef committed:
Skip re_is_literal()'s parsing an AST construction for literals.

This should find the same result anyway, there's just no need to go to the trouble of constructing an AST here, when we already know the pattern is a literal.

katef committed:
Override dialect by file extension.

katef committed:
-p for prefix.

katef committed:
-F for re_flags.

katef committed:
No need to handle endids during fsm_minimise here.

katef committed:
Don't call intersect_charset() with an empty charset, it's clearer to handle in the caller.

katef committed:
Comment generated code

katef committed:
Codegen for various languages.

I'm not particularly thrilled about the handling for different AMBIG_ modes here. I'm thinking eventually it might make sense to move this stuff into libfsm proper and share it with the other cli tools.

katef committed:
-x for unanchored literals.

katef committed:
First cut at introducing llvm ir codegen.

katef committed:
-opaque-pointers for clang 14 in github CI.

katef committed:
Include num_re_errors in per-file failures.

katef committed:
Skip an IO API/lang combination that we don't support.

katef committed:
Quieten GNU ld about something I don't understand.

katef committed:
Merge pull request #480 from katef/kate/retest-errors

Missing error handling for retest(1)

katef committed:
Maintain n index from 0, not -1.

This produces a little more code (there's an additional f: branch), but also it works. So who's judging?

katef committed:
We need -mllvm for opaque pointers as well as -opaque-pointers.

katef committed:
Run these retest tests for llvm too.

katef committed:
Consistency for print_label()

katef committed:
Factor out print_branch()

katef committed:
Centralise `fail:` nodes.

katef committed:
Whitespace.

katef committed:
Construct retlist and output deduped rets.

This isn't used yet, but it's a step towards generating phi.

This looks like:
```
fail:
	ret i32 -1
ret0:
	ret i32 u0x1 ; "a"
ret1:
	ret i32 u0x2 ; "b"
ret2:
	ret i32 u0x4 ; "c"
ret3:
	ret i32 u0x3 ; "a", "b"
ret5:
	ret i32 u0x6 ; "b", "c"
ret6:
	ret i32 u0x7 ; "a", "b", "c"
```

WIP

katef committed:
Sort endids written out by fsm_endid_get().

This is for caller convenience, I keep calling qsort for various reasons whenever I call this function, for example to use memcmp on ids.

But the final motivation here was that I wanted to memcmp a const-qualified pointer to an id set stored elsewhere, and I can't do that without allocating for a local copy. And that seems silly when we can just qsort here.

katef committed:
Merge pull request #482 from katef/kate/sort-endids

Sort endids written out by fsm_endid_get()

katef committed:
Branch to centralised retlist labels.

katef committed:
Jump via ret*: labels to a single stop:

This keeps the phi node constant size wrt the number of basic blocks.

I'm very proud of this idea.

katef committed:
An arguably less worse workaround to keep .endleaf_opaque available.

katef committed:
Merge pull request #483 from katef/kate/llvm

Add llvm codegen

katef committed:
Centralise fsm_print() and friends.

Now you call fsm_print(f, fsm, FSM_PRINT_*) rather than fsm_print_*().

This might look a bit cumbersome to the caller, but I'm doing it so we have a single entry point to put shared stuff. More on that later.

katef committed:
Missing targets.

katef committed:
Hoist up ir and vm opcode compilation to fsm_print().

This cuts down a lot of repetition for the various print functions, and also helps make clear what they actually use.

katef committed:
Return bool for match/failure from generated code.

This drops information about which endids match. My intention is to output that later, but I wanted to think about one thing at a time, because this gets confusing. Here I'm making things worse before they get better.

This applies for the codegen that doesn't use the endleaf callbacks. I'm going to call these "default" output.

One observation here is that not all FSMs will have endids attached, and we still need to indicate success/failure. So that's why I'm not using the populated endids to indicate a match.

katef committed:
No need for ir_state here.

I've carried over only the fields we actually use to the vm ir struct. This is a bit brutal, I'm trying to avoid sacrificing functionality, but when it comes to a choice, at the moment I care more about keeping things separate than not cutting features. However the only thing I think I dropped is the state numbering for -l vmdot output.

katef committed:
Introduce ambig enum, and replace leaf/endleaf callbacks with reject/accept hooks.

The accept callback gets a set of end ids. These hooks default to outputting those ids verbatim, you needn't override them per-program. I think that's clearer for users of the API.

I ported the ambig enum over from (unmerged) work on rx(1). This is for how to handle endid ambiguities, which does something appropriate to each language's generated API. So that's now built in to libfsm rather than being implemented per-language per-program.

This gets us clearer handling for multiple id output per language, and I think it's also clearer about what's responsible for what, overall.

katef committed:
Only qsort a non-empty retlist.

katef committed:
Cruft.

katef committed:
Move alloc hooks to struct fsm.

This is a step towards moving fsm_options to the print interfaces only.

katef committed:
Uninitialized value.

katef committed:
Remove fsm.opt and split print hooks from options.

The important part of the diff here is just:
```diff
--- a/src/libfsm/internal.h
+++ b/src/libfsm/internal.h
@@ -11,7 +11,6 @@
 #include <stdlib.h>

 #include <fsm/fsm.h>
-#include <fsm/options.h>

 #include <adt/common.h>

@@ -76,7 +75,6 @@ struct fsm {

        struct fsm_capture_info *capture_info;
        struct endid_info *endid_info;
-       const struct fsm_options *opt;
 };
```

Everything else is fallout from not needing to pass around the options struct. This leaves the options struct only used for the print routines.

katef committed:
fsm_dump(), a convenience to replace the original fsm_print() for debugging.

Now fsm_print() has a bunch of arguments, it's going to be annoying for debugging during development. fsm_dump() returns the same interface, but with some options set for compact output.

The output looks like this:

```
# src/re/main.c:1081 new
0; 1;

0  ->  1 "a" .. "c", "x"; # e.g. "a"

start: 0;
end:   1;
```

katef committed:
Port CLI tools to use opt.ambig and the conflict callback.

There are a couple of awkward spots here (in particular calling fsm_endid_get() seems cumbersome for a user), but overall I think this came out really nicely. It does simplify a lot of the caller-side bookkeeping around tracking conflicts. And that's reflected in the diff removing program-specific datastructures.

katef committed:
Clarification.

katef committed:
-s for overriding dialect by file extension.

katef committed:
-e and -E for prefixes.

katef committed:
Move out retlist construction.

katef committed:
Constuct retlist ahead of time for all vm-based formats.

Now the VM IR ops point to retlist entries, rather than carrying their own endid sets. This means we can use the same indexing to de-dup them for every VM-based format.

katef committed:
Add AMBIG_MULTIPLE output for go.

The generated code looks like this:
```go
package fsm_fsm

var ret0 []uint = []uint{1}
var ret1 []uint = []uint{2}
var ret2 []uint = []uint{1, 2}
var ret3 []uint = []uint{0, 1, 2}

func fsm_Match(data string) (bool, []uint) {
	var idx = ^uint(0)

	if idx++; idx >= uint(len(data)) {
		return true, ret0
	}

	if data[idx] == 'a' {
		goto l3
	}

	if data[idx] == 'b' {
		goto l2
	}

	if data[idx] != 'c' {
		return false, nil
	}

l0: // e.g. "c"
	if idx++; idx >= uint(len(data)) {
		return true, ret2
	}

	if data[idx] <= '`' {
		return false, nil
	}

	if data[idx] <= 'b' {
		goto l1
	}

	if data[idx] == 'c' {
		goto l0
	}

	{
		return false, nil
	}

l1: // e.g. "aa"
	if idx++; idx >= uint(len(data)) {
		return true, ret1
	}

	if data[idx] <= '`' {
		return false, nil
	}

	if data[idx] <= 'c' {
		goto l1
	}

	{
		return false, nil
	}

l2: // e.g. "b"
	if idx++; idx >= uint(len(data)) {
		return true, ret3
	}

	if data[idx] == 'a' {
		goto l1
	}

	if data[idx] == 'b' {
		goto l2
	}

	if data[idx] == 'c' {
		goto l1
	}

	{
		return false, nil
	}

l3: // e.g. "a"
	if idx++; idx >= uint(len(data)) {
		return true, ret1
	}

	if data[idx] == 'a' {
		goto l1
	}

	if data[idx] == 'b' {
		goto l2
	}

	if data[idx] == 'c' {
		goto l1
	}

	{
		return false, nil
	}

}
```

katef committed:
No need for the vmops_dialect enum here.

katef committed:
Whitespace.

katef committed:
Bugfix; 1 means VM_END_SUCC here.

katef committed:
Support for enids for the vmops output.

I've handled AMBIG_NONE here, but I haven't distinguished the other ambig modes. Other than AMBIG_NONE, the other modes are all presented as an array of ids, even if it's just a single element. That's because I don't see any reason to give these specialised APIs for the current use-cases for this generated code, which is supposed to be a direct representation of our VM opcodes.

katef committed:
Ensure we don't pass NULL to memcmp()

katef committed:
Typo.

katef committed:
Merge pull request #485 from katef/kate/ambig-mode

Introduce ambig mode

katef committed:
First cut at AMBIG_MULTIPLE for llvm.

It's a bit rough, this isn't what I want to end up with, but I wanted to
commit this as a waypoint.

katef committed:
Rework stop: to index into an array for return values.

This means the phi instruction now only carries an array index.

Many thanks to @mcy for advice and patient help here.

katef committed:
Don't use `undef` for values the caller is expected to not access, use `poison`.

Thanks to @mcy for this.

katef committed:
Split overriding comments to a separate hook.

This allows callers to default the codegen for accepting states (in particular outputting the values for endids) independently of commenting caller-specific meanings for the IDs.

katef committed:
No need to emit a bitmap here.

Originally I'd intended this as a demonstration of how various applications can handle ambiguities differently. But now we have library support for AMBIG_MULTIPLE, I think this is just confusing.

katef committed:
Use id count to indicate match failure, and >= 0 means success.

The idea here is just to trim down %rt to:
```
%rt = type { ptr, i64 }
```

where we clearly don't need a uint64_t's count of unique ids.

I've purposefully not done the same for single-id interfaces. I don't want to mix success/failure with an id *value*, because the values are opaque (i.e. the meaning of an id value is the responsibility of the caller). Whereas here for AMBIG_MULTIPLE I'm mixing success/failure with the count, not with the id values.

Suggested by @mcy, thank you.

katef committed:
Clarification.

katef committed:
Add AMBIG_MULTIPLE output for Rust.

The generated code looks like this:
```rust
pub fn fsm_main(input: &str) -> Option<&'static [u32]> {
    use Label::*;
    static RET0: [u32; 1] = [1];
    static RET1: [u32; 1] = [2];
    static RET2: [u32; 3] = [0, 1, 2];

    let mut bytes = input.bytes();

    pub enum Label {
        Ls, L0,
    }

    let mut l = Ls;

    loop {
        match l {
            Ls => { // e.g. ""
                let c = match bytes.next() {
                    None => return Some(&RET0) /* "x?" */,
                    Some(c) => c,
                };
                if c != b'x' { return None }
                let c = match bytes.next() {
                    None => return Some(&RET2) /* "x", "x?", "x+" */,
                    Some(c) => c,
                };
                if c != b'x' { return None }
                l = L0; continue;
            }

            L0 => { // e.g. "xx"
                let c = match bytes.next() {
                    None => return Some(&RET1) /* "x+" */,
                    Some(c) => c,
                };
                if c != b'x' { return None }
                l = L0; continue;
            }
        }
    }
}
```

katef committed:
Factor out print_ret().

katef committed:
Merge branch 'main' into kate/more-multi

katef committed:
Merge branch 'kate/more-multi' into kate/rx

katef committed:
Whoops... wrong exit status.

katef committed:
Oops... this should've been done with the API change to return a bool.

katef committed:
Separate calling points for the comments hook.

This gives better control over whitespace and punctuation between the hooks. For example we can output "<accept>, <comment>\n" with a comma between, and that sits more nicely for single-line comments. Previously these had to be "<accept> <comment>,\n"

katef committed:
Typo.

Spotted by Scott.

katef committed:
Merge pull request #486 from katef/kate/more-multi

Implement AMBIG_MULTIPLE for llvm, Rust, go, and the vmops dump

katef committed:
Merge branch 'main' into kate/rx

katef committed:
-t for AMBIG_EARLIEST

katef committed:
Graphviz output for the DFA.

katef committed:
-a and -w for anonymous states and fragment output.

katef committed:
-X for always_hex

katef committed:
-Fb for consistency wrt -Fi.

katef committed:
Documentation for -F flags and -k io api.

katef committed:
A little more progress on invocation.

katef committed:
Merge pull request #487 from katef/kate/man-flags

Document -F <flags> and -k <io>

katef committed:
Manpage for rx(1)

katef committed:
Missing flag, this is now present on main.

katef committed:
Stray assertion.

I'm not sure why I thought this was neccessary; I tested and the trie code seems happy with an empty string. Which is fortunate, because I'm a believer in recursive datastructures.

katef committed:
Spelling.

sw17ch committed:
In `closure_free()`, use the FSM allocator instead of `free()`.

katef committed:
Merge pull request #489 from sw17ch/sw17ch/fix-closure-free-alloc

In `closure_free()`, use the FSM allocator instead of `free()`.

katef committed:
Wrong exit status for -G.

Spotted by @pierreganty (#478), thank you

katef committed:
Wrong return type for AMBIG_ERROR/AMBIG_EARLIEST.

Spotted by @sw17ch, thank you.

katef committed:
-C opt-level=3 makes a huge difference here.

Spotted by @sw17ch, thank you.

katef committed:
Missing language.

katef committed:
Off by one.

Previously this gave:
```
; ./build/bin/re -r literal -G 3 abc
; ./build/bin/re -r literal -G 4 abc
abc
;
```
and now -G 3 does construct "abc" (not including the newline, which is a property of the printing, not of the constructed string):
```
; ./build/bin/re -r literal -G 3 abc
abc
; ./build/bin/re -r literal -G 4 abc
abc
;
```

and:
```
; ./build/bin/re -r native -G 3 '^x+$'
x
xx
xxx
```

Spotted by @pierreganty (#478), thank you

katef committed:
Stray assertion.

This doesn't help for #317, but whatever the solution is there, asserting about it is the wrong thing to do.

Spotted by @classabbyamp, thank you

katef committed:
Bump actions for Node 20.

katef committed:
Happier cache lines.

Probably the only change that actually affects something, `struct ir_state` is both important and now fits within a single cache line.

I did this by eyeballing:
```
; bmake -r -j 10 CC=clang DEBUG=1 && (for t in `cat /tmp/tags`; do pahole --class_name $t build/src/*/*.o build/src/adt/*.o ; done ) > /tmp/tt
```

This gives a ~2% improvement on total execution for rx -q /usr/share/bin/words

katef committed:
Guard .tag; it helps keep structs tidy.

katef committed:
No need for calloc here.

katef committed:
Update for hooks & options API changes.

I have extremely broken the iprange example. I am very unsure what's going on with this program.

katef committed:
Merge pull request #490 from katef/kate/bugfixin

A few small bugfixes

katef committed:
Spelling.

Spotted by Dan Kegel, thank you.

silentbicycle committed:
experimental: Add eager outputs, similar to endids but eagerly matched.

When combining several unanchored regexes it becomes VERY expensive
to handle combinations of matches via the end state -- essentially,
the whole reachable DFA gets separate matching and non-matching copies
for each pattern, leading to a DFA whose size is proportional to the
number of *possible combinations* of matches. With eager outputs,
we can set a flag for matching as we reach the end of the original
pattern (before looping back and possibly also matching other patterns),
which keeps the state count from blowing up in fsm_determinise.

To see how much difference this makes, the test tests/eager_output/run7
combines 26 different patterns. It should finish very quickly (~50 msec,
just now). Try running it with `env FORCE_ENDIDS=N` for N increasing from
4 to 26. Around 10-11 it will start taking several seconds, and memory
usage will roughly double with each step.

This PR adds `fsm_union_repeated_pattern_group`, a variant of
`fsm_union_array` that combines a set of DFAs into a single NFA, but
correctly handles a mix of anchored and unanchored ends without the
state count blowing up. It currently needs flags passed in for each fsm
indicating whether the start and/or end are anchored, and there is a
hacky special case that removes self-edges from states with eager
outputs and instead connects them to a single overall unanchored end
loop. I haven't yet figured out how to handle this properly in the
general case, but it works for this specific use case, provided all the
DFAs are combined at once. (Combining multiple DFAs each produced by
determinising fsm_union_repeated_pattern_group's result probably won't
work correctly.) I have tried detecting and ignoring those edges in
fsm_determinise, after epsilon removal, but so far either it still
causes the graph size to blow up or subtly breaks something else.

This is still experimental, and the code generation for `-lc` here is
quite hacky -- it expects the caller to define a `FSM_SET_EAGER_OUTPUT`
macro, since the code generation interface doesn't define where the
match info will go yet. A later PR will add a new code generation mode
with better support for eager outputs, and I plan to eventually
integrate this better with rx, AMBIG_MULTIPLE, and so on.

(This squashes down a couple false starts.)

katef committed:
Merge pull request #488 from katef/kate/rx

rx, a program for compiling sets of regular expressions

katef committed:
Merge pull request #491 from katef/kate/happy-cache-lines

Happier cache lines

silentbicycle committed:
trim's integrity_check should only run with EXPENSIVE_CHECKS.

silentbicycle committed:
determinise: Remove stale comment.

silentbicycle committed:
common.h: TRACK_TIMES should be 0 when BUILD_FOR_FUZZER is set.

It fills up the fuzzer logs with lots of uninteresting output.

silentbicycle committed:
bugfix: fsm_endid_get should sort with result count, not buffer size.

fsm_endid_get's id_buf_count argument is expected to "have enough
cells (according to id_buf_count)", but if it has more than enough,
stale data can get sorted into the result.

Add a test, tests/endids/endids_reused_buffer.c

silentbicycle committed:
endids_reused_buffer.c: Fix memory leak in test. (free fsm.)

katef committed:
Merge pull request #493 from katef/sv/bugfix-endid-qsort-count

bugfix: fsm_endid_get should sort with result count, not buffer size.

silentbicycle committed:
Add re_is_anchored interface.

data-man committed:
Update to Unicode 16.0

silentbicycle committed:
fsm_compact_states must remap endids, to avoid dangling references.

Add a regression test showing possible endid false negatives when FSMs were trimmed
(called from fsm_minimise) without updating endids.

katef committed:
Merge pull request #495 from data-man/ucd16

Update to Unicode 16.0

silentbicycle committed:
Change fsm_detect_required_characters interface.

`struct bm` isn't part of the public API, so use a uint64_t[4]
instead, and add an optional parameter for the count.

Update the tests.

silentbicycle committed:
fsm_detect_required_characters: Set count to 0 when matching "".

silentbicycle committed:
Move detect_required1.c tests to .re/.txt files. Add -q chars to fsm.

silentbicycle committed:
fsm.1.xml: Add basic info about -S <step_limit> and -q requiredchars.

silentbicycle committed:
print API: Box end_ids and end_id_count in a struct for callbacks.

Move the end_id array and its count into a struct for state metadata,
and rename access throughout to end_ids and end_id_count.

Upcoming changes for eager output IDs will soon be passing more info to
all of these callbacks, but only callers making use of those fields need
to care. Instead of making callers add more `(void) param;` declarations
all over the place to avoid warnings, just pass in a metadata struct
pointer. Also, "count" is a pretty generic name and what it refers to
will soon be ambiguous.

This should not be a functional change on its own.

silentbicycle committed:
Move call to fsm_countstates inside #ifndef NDEBUG.

It's only used for the assertion.

silentbicycle committed:
libfsm.syms: Expose fsm_new_statealloc.

It's already in fsm/fsm.h.

silentbicycle committed:
ir: Wrap ids and count in a struct, to namespace them as endids.

The IR struct is about to get another id & count pair.

This loses storing the count as a 31-bit bitfield, but if the
goal for that is saving memory then the ids array allocation could
be replaced with a struct that contains the count, and then each
IR without endids will save more space than the current approach.

katef committed:
Merge pull request #492 from katef/sv/determine-which-characters-must-appear-in-input-to-match

Add fsm_detect_required_characters

katef committed:
Merge pull request #496 from katef/sv/bugfix-compact-states-should-remap-endids

fsm_compact_states must remap endids, to avoid dangling references.

katef committed:
Merge pull request #497 from katef/sv/move-codegen-callback-metadata-args-into-struct

print API: Box end_ids and end_id_count in a struct for callbacks.

silentbicycle committed:
fsm_generate_matches: Add seed argument, random choice if non-zero.

There isn't a flag to set the seed in fsm or re yet.

Update callers, add a 0 seed argument to default to the
existing behavior.

Also, no end states means nothing to do, so exit right away.

silentbicycle committed:
Temporary interface: Add fsm_determinise_with_config.

This works, but we may want to make some naming changes. I'd prefer to
use a struct rather than a single parameter here -- as long as other
fields have a sensible default when zeroed, the callers will only need
to make the interface change once.

silentbicycle committed:
Add 'default:' (should be unreachable) for switch case.

This is getting rejected by CI.

silentbicycle committed:
Ensure .has_eager_outputs is zeroed on new states. (msan)

silentbicycle committed:
eager_output interface cleanup: Replace _any with _count and _get.

fsm_eager_output_count and fsm_eager_output_get aligns better with
the endid interface, and now  fsm_eager_output_get ensures the buffer
contents are sorted,

silentbicycle committed:
minimise_test_oracle.c: mismatched eager outputs also prevent merging.

Do the same check with eager outputs as it does with endids -- any
two states with distinct sets for either should end up in different
equivalence classes.

katef committed:
Merge pull request #498 from katef/sv/determinise-with-config

Temporary interface: Add fsm_determinise_with_config.

silentbicycle committed:
fuzzer: Add seed argument for fsm_generate_matches (interface change).

silentbicycle committed:
Remove direct call to srand(), just recommend the caller use it.

Instead of taking a seed argument, just treat `int randomized`
like a flag. (I'd use `bool`, but the header consistently uses
`int` for boolean arguments.)

katef committed:
Merge pull request #499 from katef/sv/randomized-generation-and-misc-fixes

Randomized generation and misc. fixes

katef committed:
Merge pull request #500 from katef/sv/re-is-anchored

Add re_is_anchored interface.

silentbicycle committed:
fuzz/target.c: re_is_anchor interface changes.

katef committed:
Missing documentation for -x

katef committed:
Too much documentation for -u.

katef committed:
Merge pull request #502 from katef/kate/missing-manpage-flag

Missing documentation for -x in the rx(1) manpage

silentbicycle committed:
Fix memory leak.

The edge sets leak when halting with FSM_DETERMINISE_WITH_CONFIG_STATE_LIMIT_REACHED.

silentbicycle committed:
Add a regression test for the memory leak.

I verified that LeakSanitizer catches the leak during this test when the
fix is reverted.

This commit adds a test directory, `tests/regressions`, for misc.
regression tests that need a .c file and aren't easily testable via fsm,
re, etc. inputs.

katef committed:
Merge pull request #504 from katef/sv/fix-memory-leak-from-determinise-with-config

Fix a memory leak during `fsm_determinise_with_config`'s early exit

silentbicycle committed:
Fix memory leak in fsm_eager_output_compact, found while fuzzing.

silentbicycle committed:
Fix fsm_union_repeated_pattern_group's anchoring linkage.

Previously this didn't handle mixed anchoring correctly, potentially
leading to false positives the case represented by
eager_output_alt_mixing_anchored_and_unanchored.c.

See comments in fsm_union_repeated_pattern_group for details.

Fuzzing did not turn up any new issues.

Another commit after this will make a few small interface changes
and update callers.

silentbicycle committed:
fsm_union_repeated_pattern_group: Interface changes.

- Instead of taking an array of `struct fsm_union_array *` pointers,
  this now takes an array of `struct fsm *` pointers. The other
  fields on `fsm_union_array` are no longer used, so the extra
  struct layer has been removed.

- This now takes an extra argument, id_base, because each
  nfa[i] will get end IDs and/or output IDs (i + id_base) set on them.
  Previously these were set by the caller.

- Rename parameters, to emphasize that the FSMs must be NFAs.

- Update the test code for the interface changes.

- Remove flags from the test code that are no longer used.

silentbicycle committed:
fsm_union_repeated_pattern_group: fix linkage for mixed start anchoring.

deg4uss3r committed:
Update CI to use Ubuntu 22.04 for now so we do not get bleeding edge clang and fail builds

silentbicycle committed:
Add tests, fix anchoring bugs in fsm_union_repeated_pattern_group.

There are two bugs captured in eager_output_unanchored_end_plus.c:

- Regexes ending in '+' weren't combining correctly, because analysis
  wasn't properly handling the construction for matching but optionally
  repeating the last character.

- Eager matching after consuming a single character from the start state
  wasn't linked correctly to the global_unanchored_start_loop, so while
  the labeled edges were copied the eager output was lost.

The other test files are focused on variants of that -- the + and
start cases individually, and when + precedes a `()` subtree with
more than one character.

silentbicycle committed:
Interface change: Add 'const'.

silentbicycle committed:
union: Fix trivial memory leak.

silentbicycle committed:
union.c: Add comments for assertions.

Fuzzing has produced inputs that cause this to fail, but they
all depend on embedded '\0' characters. I wasn't able to reproduce
the failure without those present, but I will investigate further
later. For now, adding a TODO.

silentbicycle committed:
Switch to collecting an anchored_start state set, not just one state.

`(^|wax-)((?:banana|^apple))` is an example of a regex that needs
multiple anchored_start states linked in order to combine correctly.

katef committed:
Merge pull request #505 from deg4uss3r/rth/fix-ci

Update CI to use Ubuntu 22.04

silentbicycle committed:
retest: Remove `isatty` check and extra logging output.

While discussing other issues, kate said she'd rather not have this
have different behavior when stdout is / isn't a tty.

katef committed:
Merge pull request #506 from katef/sv/remove-isatty-check

retest: Remove `isatty` check and extra logging output.

silentbicycle committed:
Remove dependence on internal.h in print/ir.h and lx/print/c.c.

print/ir.h only depended on it for FSM_SIGMA_COUNT in `struct
ir_state_table`, and IR_TABLE isn't actually implemented, so
this can be removed for now.

lx/print/c.c only needed internal.h because of print/ir.h, and
because of several direct accesses to fsm->statecount, which are
easily replaced by calls to fsm_countstates.

silentbicycle committed:
Fix lx by rewriting much of its -l c codegen.

Changing the leaf and endleaf callbacks to accept and reject in #485
broke lx, but it went unnoticed for a while. This fixes it.

libfsm's normal execution mode evaluates a DFA, character by character,
terminating either when the next character isn't a valid edge or the end
of input is reached (in which case it checks end state metadata). lx's
execution mode is a little different, because it's tokenizing -- instead
of reading to the end of input, it should consume as much consecutive
input that matches a particular token, then push back the last character
read (so it can resume with it as context for the next token), yield the
token type, and suspend.

lx used to work by breaking abstraction and calling directly into
`fsm_print_cfrag` (overriding the leaf behavior to yield token types,
and adding an extra 'NONE' state to the generated state machine code),
but when the callback interfaces shifted its internals no longer fit
what lx expected. Now the reject hook is passed the same state
metadata as the accept state, and the reject hook in lx checks whether
the end id is associated with a particular AST mapping and token type.

This is only implemented for the "c" output format, but similar changes
could possibly make others usable without a lot more work. In
particular, kate mentioned it'd be good to be able to use "vmc" output
instead of "c" moving forward.

Most of the code changes happen inside of lx's code generation, but
there are a few elsewhere:

- The reject hook now has a state_metadata pointer, so update the
callers for all the output formats.

- libfsm's 'c' output now includes a macro `FSM_ADVANCE_HOOK(C)`,
which is called with the next character read in the FSM_IO_STR and
FSM_IO_PAIR io modes immediately after advancing. This is used to inform
lx's internal bookkeeping about token positions and buffering token
names. FSM_IO_GETC doesn't need it, because its getc callback manages
the character stream. The macro defaults to a no-op when undefined.

- libfsm's 'c' output also includes a flag, `has_consumed_input`, so the
code expanded in place from the reject/accept hooks can determine when
the state machine input handler loop has consumed any input. This was
previously encoded by the extra NONE state.

lx's code generation using this flag is a bit cluttered, because the
reject hook doesn't know whether it's expanding for the end states, but
it's probably not worth changing the reject hook type signature to add
another flag. This results in checks for has_consumed_input in code
paths where trivial static analysis would show it to be dead code, and
some extra unreachable code at the end of the function.

silentbicycle committed:
Makefile: Check '*res*' not 'res*' for tests.

This will miss failures in prefixed res files, such as

    build/tests/lxpos/dyn-fdgetc-getc-res0

silentbicycle committed:
test/lxpos: Update test data (EOF position info).

Instead of having the EOF token occupy the same byte, line, and column
position as the last token, it should immediately follow.

The new lx codegen behaves this way, and katef and I decided that it
made sense to keep it like that, as long as it's consistent.

silentbicycle committed:
Re-enable lxpos tests.

Add `${LX}` as a dependency for the targets using it.

Remove the `getcio=${io}` and `io=${io}` arguments to cat. Those may be
a merge error? They just produce a warning.

silentbicycle committed:
Use $LX_BIN instead of $LX in lxpos makefile.

Some of the CI test matrix builds set LX to 'true; echo lx', but that
obviously won't work for tests that actually need to run lx in order to
exercise its output.

silentbicycle committed:
lx: Make -l dump's output call lx.free() when using dynamic buffer.

silentbicycle committed:
lx: Use prefix.tok, not "TOK_".

silentbicycle committed:
lx: return TOK_ERROR if reaching the end of a zone function.

silentbicycle committed:
lx: Handle unexpected EOF in pattern pairs. Add tests, update out6.dump.

Add a couple test cases (in8-10.txt) with an unexpected end of input,
either in the middle of a pattern, or after matching the first pattern
in a .. pair, but without matching the second.

Supporting this changes the expected result for in6.txt: Previously it
resulted in TOK_EOF, now it leads to TOK_UNKNOWN and produces a
"lexically uncategorised" error message for the unexpected end of input.
This change is necessary for fixing #386 / #508, and more generally
to detect things like unterminated string literals.

silentbicycle committed:
lx: Rewrite logic to make the four cases explicit, fix dead code.

This was previously ending up with a useless call to the current zone
after returning the token ("case S1: return TOK_UNKNOWN; lx->z(lx);"),
which led to a warning in CI [-Werror=implicit-fallthrough=].

silentbicycle committed:
lx: Only gen fixedpop / dynpop & calls to them when buffer mode is set.

silentbicycle committed:
lx: Suppress warning for possibly unused function.

These may or may not be called, depending on the input.

silentbicycle committed:
lx: Ensure prefix.api & prefix.lx are used in the generated code.

In some cases this was hardcoding "lx_" in the generated code, which
could lead to build failures if 'lx -e' was used to override the default
prefix.

silentbicycle committed:
Replace FSM_ADVANCE_HOOK macro with optional hooks->advance callback.

silentbicycle committed:
The advance hook should also be called for FSM_IO_STR.

silentbicycle committed:
Move setting `has_consumed_input` flag into lx's advance hook.

This avoids cluttering libfsm's print output with `has_consumed_input`,
which is specific to lx.

silentbicycle committed:
lx: Avoid useless call to pop and some other 'unused' warnings.

Tested with every combination of (dyn+fgetc, fixed+fgetc, pair, str),
with and without '-x buf', '-x pos', or both.

katef committed:
Merge pull request #509 from katef/sv/fix-lx-token-identification

Fix lx token identification

silentbicycle committed:
lx: Distinguish between unexpected EOF and EOF in ignored zones.

PR #509 introduced a bug: It didn't distinguish between an unexpected
end of input and an end of input in a zone that matches but ignores its
input. This caused several lxpos tests to fail due to getting a
TOK_UNKNOWN rather than a TOK_EOF when the input has trailing
whitespace, but I didn't notice until after merging because the normal
build doesn't regenerate the code for src/lx/lexer.lx or
src/libfsm/lexer.lx. (I had ensured all the libre dialect lexers and
parsers were regenerated, but missed those.)

Instead of always printing TOK_UNKNOWN, this this inspects the zone
mappings to determine whether the current end ID represents a dead end
for the zone. If not, it should instead print TOK_EOF.

silentbicycle committed:
Generated code. Re-generate lexers and parsers with lx bug fixed.

katef committed:
Typo.

katef committed:
Stray const.

katef committed:
Merge pull request #511 from katef/kate/stray-const

Stray const in generated code

katef committed:
Merge pull request #510 from katef/sv/fix-lx-handling-for-EOF-broken-by-509

lx: Distinguish between unexpected EOF and EOF in ignored zones (broken by #509)

silentbicycle committed:
union.c: Updates to fsm_union_repeated_pattern_group and its internals.

- Remove some dead code.

- Update several comments.

- Rename a couple functions to clarify their use.

- Refine detection of the unanchored start loop.

data-man committed:
Update to Unicode 17.0

silentbicycle committed:
Eager outputs: Comment on use, rename some functions, clean up.

silentbicycle committed:
Rename test file -- make tests looks for build files matching "*res*".

`build/tests/eager_output/run_mixed_start_anchor_regression` matches
"*res*" because of "regression", so `make test` incorrectly treats it
like any of the other output test result files. Rename it, to remove
"regression".

This is an unexpected consequence of cb42d58f.

silentbicycle committed:
fuzz/target.c: Updates for interface changes.

silentbicycle committed:
Misc. cleanup before integration.

silentbicycle committed:
Restore FORCE_ENDIDS behavior for tests/eager_output/eager_output7.c.

This was removed in af27a87c, while updating for interface changes, but
it's worth keeping this. It makes the performance difference between
using endids and eager outputs for combining larger sets of unanchored
patterns obvious -- on my current machine,
`build/tests/eager_output/run7` completes in about 55 msec, whereas `env
FORCE_ENDIDS=7 build/tests/eager_output/run7` takes about 88 msec, 8
takes 200 msec, 9 takes 639 msec, 10 takes 1.62 seconds, 11 takes 3.9
seconds, ... without FORCE_ENDIDS, it's running all 26 patterns.

(FORCE_ENDIDS < 7 finishes faster, but that's because it's only using
the first couple patterns and combining into a much smaller DFA.)

Also, a minor bugfix -- instead of asserting that the outputs were >=
the expected ID count, check that it's exactly equal. This was due to
IDs set by both eager outputs and endids being counted twice; some of the
eager_output_alt_mixing_anchored_and_unanchored.c cases could lead to
this, due to the mixing of anchored and unanchored subtrees in the
regexes.

silentbicycle committed:
fsm.h: Fix missing word in comment.

silentbicycle committed:
Note why fsm_union_repeated_pattern_group depends on re_comp.

This may not be necessary in the future, but currently analyze_group_nfa
in union.c depends on knowing that re_comp will always assign the true
unanchored start loop an earlier state ID than the unanchored end loop
in order to properly link a few ambiguous cases.

katef committed:
Merge pull request #512 from data-man/ucd17

Update to Unicode 17.0

silentbicycle committed:
Optionally save linkage_info during NFA construction in re_comp.

This adds a new flag for re_comp, RE_SAVE_LINKAGE_INFO. If that's set,
then save information about the unanchored start and end loops, the
end loop end, anchored starts, and anchored ends while building the
NFA from the regex. Previously, this info was recovered by analyzing the
NFA structure inside of fsm_union_repeated_pattern_group, but it seems
less brittle to save those details during construction.

This commit adds the flag, saves the info during construction, and
after doing analysis it checks the results match. Since there is one
case where it doesn't (eager_output_fr3.c), print a message and override
it. This was actually a bug: that particular edge case leads to
detecting the same state as both the unanchored start and end loops. In
that specific case, it seems to be harmless, but this commit also adds
an assert that the USL and UEL are not the same state.

From now on, fsm_union_repeated_pattern_group depends on all of the NFAs
passed to it having the linkage_info struct populated. They should come
fresh from re_comp, without determinising or other transformations. I
added an assert at the end of determinisation checking that the DFA does
not have a linkage_info struct.

Note: fsm_clone currently doesn't copy the linkage_info. Maybe it should.

silentbicycle committed:
Merge branch 'main' into sv/eager-outputs-and-union-repeated-pattern-group

katef committed:
Stray comment.

silentbicycle committed:
Copy some fields from linkage_info, remove analysis.

They need to be copied over (transferring ownership of the state sets)
because the linkage_info struct gets freed during the call to fsm_merge.

This removes some code that tended to reconstruct the same information
by analyzing the NFA structure, but was significantly more compliacted,
and in at least one case it led to the wrong result (albeit harmlessly).
It's simpler to just save the relevant state IDs during NFA construction.

silentbicycle committed:
Restore freeing of state sets copied from linkage_info.

silentbicycle committed:
Update stale comment.

This wasn't updated to mention RE_SAVE_LINKAGE_INFO.

katef committed:
Merge pull request #513 from katef/sv/eager-outputs-and-union-repeated-pattern-group

Eager outputs and `fsm_union_repeated_pattern_group`

gsusanto-fastly committed:
Add documentation guide on how to use Libfsm effectively

gsusanto-fastly committed:
Update docs based on reviews

gsusanto-fastly committed:
Revision #2

katef committed:
Naming.

katef committed:
Markup.

katef committed:
Blurb on calling the generated code.

katef committed:
Blurb on bounded repetition.

(Contributed by Scott)

katef committed:
Markup.

katef committed:
Merge pull request #515 from katef/gsusanto-docs

Advice on how to use libfsm for generating performant pattern matchers

katef committed:
First cut at re_interpolate_groups()

katef committed:
Add start,end error reporting

katef committed:
Clarification.

katef committed:
Fill out placeholders for writing out output.

katef committed:
Convincing myself string offsets are convenient

katef committed:
Allow a NULL output string.

katef committed:
Clarification.

Spotted by both June and Scott, thanks

katef committed:
Defensively terminate the output buffer on error.

Suggested by Scott

katef committed:
fail-on-cache-miss: for grabbing arbitrary builds.

katef committed:
Update to actions/cache@v5

katef committed:
cache/restore where possible.

katef committed:
Explicitly allow build cache miss for makefile tests.

katef committed:
Explicitly fail-on-cache-miss for other things too.

katef committed:
Merge branch 'kate/actions-fluffery' into kate/interpolate_groups

katef committed:
Merge pull request #517 from katef/kate/actions-fluffery

Github actions cache fluffery

katef committed:
Merge pull request #516 from katef/kate/interpolate_groups

Add re_interpolate groups()

katef committed:
Merge mishap, accidentally @v4

silentbicycle committed:
Makefile: grep for 'FAIL' should use -I to ignore binary files.

I ran into a test failure:

    grep FAIL build/tests/*/*res*; [ $? -ne 0 ]
    grep: build/tests/eager_output/run_mixed_start_anchor_regression: binary file matches
    *** Error code 1

because tests/eager_output/eager_output_mixed_start_anchor_regression.c
contains the substring "res" in "regression", and the binary file
contains "FAIL:

    $ strings build/tests/eager_output/run_mixed_start_anchor_regression | grep FAIL
    VM_END_FAIL
    VM_FAIL

This shouldn't matter for testing purposes, the 'test' target only cares
about the test result files containing "PASS" or "FAIL". The change in
cb42d58 to allow prefixed result files (e.g. "dyn-fdgetc-getc-res0")
made this check ANY files with "res" in the name, but grep should ignore
binary files.

katef committed:
Merge pull request #519 from katef/sv/grep-test-ignore-binary-files

Makefile: grep for 'FAIL' should use -I to ignore binary files.

katef committed:
Oops... missed installing <re/groups.h>

silentbicycle committed:
SEE LOG. Merge branch 'sv/integration-target--combinable-DFA-capture-resolution' into sv/integrate-combinable-DFA-capture-resolution

This was a really gnarly merge and took a while to get through.

Both the branch implementing captures and the branch implementing
`fsm_union_repeated_pattern_group` made interface changes and added
significant behavior to `ast_analysis.c`, in ways that were at times
tricky to reconcile. `ast_compile.c` was also restructured, but some
of the changes were necessary to inform capture and unioning
functionality.

As of now _almost_ all the tests are passing, but there are a couple
specific things to note:

- This brings in an interface change to `fsm_endid_set`, it now returns
  an enum rather than just an int. Because if you squint enough they are
  both just numbers (according to the C language spec) code using
  `if (!fsm_endid_set(...)) { ... }` will not get a warning for the
  changed meaning of the return code. I prefer having this be an enum,
  but it IS an interface change, and I'm not opposed to changing it back
  in a later commit.

- `build/tests/capture/res_test_case_list:FAIL`

  I'm going to fix this in a later commit. I'm not 100% sure yet, but
  I think it's related to conflicting changes in the parser code, which
  I'm waiting to regenerate until after this merge commit.

- `build/tests/endids/res10_minimise_partial_overlap:FAIL`

  This has to do with `AST_ANALYSIS_ERROR_UNSUPPORTED_PCRE` or
  `AST_ANALYSIS_ERROR_UNSUPPORTED_CAPTURE` not being handled yet by code
  specific to the native dialect. I'm going to handle that later, but
  will have to check whether native should behave like PCRE in those
  particular cases or not. Either the error handling code needs to be
  updated, or the code raising the UNSUPPORTED error needs to check
  which regex dialect is in effect.

- fuzz/target.c

  I have confirmed that the merged fuzzer harness code builds, but
  haven't yet spend time re-fuzzing anything. In my experience libfuzzer
  has bit-rotted over the last couple clang/LLVM releases and tends to
  nondeterministically crash in combination with some of the clang
  sanitizers now, so we may want to retarget this to using AFL-Fuzz++.
  That's well outside the scope of this PR, though.

- src/lx/parser.act

  I updated this but haven't re-generated the parsers yet, and I
  updated the generated code directly with a one-line change to
  reflect the `fsm_endid_set` interface change. As mentioned above,
  I'll re-generate that code in a separate commit.

Conflicts:
- fuzz/target.c
- include/adt/stateset.h
- include/re/re.h
- src/adt/stateset.c
- src/fsm/main.c
- src/libfsm/Makefile
- src/libfsm/capture.c
- src/libfsm/clone.c
- src/libfsm/closure.c
- src/libfsm/consolidate.c
- src/libfsm/determinise.c
- src/libfsm/determinise_internal.h
- src/libfsm/endids.c
- src/libfsm/epsilons.c
- src/libfsm/exec.c
- src/libfsm/internal.h
- src/libfsm/merge.c
- src/libfsm/minimise.c
- src/libfsm/state.c
- src/libre/ast.h
- src/libre/ast_analysis.c
- src/libre/ast_analysis.h
- src/libre/ast_compile.c
- src/libre/ast_rewrite.c
- src/libre/re.c
- src/lx/parser.act
- src/re/main.c
- tests/capture/captest.c
- tests/capture/captest.h
- tests/capture/capture3.c
- tests/capture/capture4.c
- tests/capture/capture5.c
- tests/capture/capture_concat1.c
- tests/capture/capture_concat2.c
- tests/capture/capture_union1.c
- tests/minimise/minimise_test_case_list.c

silentbicycle committed:
Regen lx parser: call re_comp with `| RE_NOCAPTURE`.

lx has its own way of handling how matches are associated with tokens,
so currently it shouldn't use captures.

silentbicycle committed:
captest.c: Instead of using `captest_getc` just use `fsm_sgetc`.

There's no need to have a custom replacement here. It also had a bug,
because it advanced the input position on EOF -- that meant that if
re_comp internally requested more characters after reaching EOF, it
would read past the end of the input string, causing tests to fail with
bizarre systax errors.

This previously wasn't an issue, but something in the merge changed
a caller's lookahead/pushback behavior so that it needs to yield EOF
forever once it reaches EOF.

silentbicycle committed:
Consolidate AST_ANALYSIS_ERROR_UNSUPPORTED_* cases and RE_EUNSUP*.

After discussing this with katef, we don't need distinct cases for when
something is unsupported because of capture behavior or because of
particular PCRE edge cases, the general unsupported error covers it.

This fixes test failures for native dialect tests, beacuse the native
dialect (understandably) doesn't handle the PCRE-specific error codes.

silentbicycle committed:
Update native parser.sid: '$' is `ast-make-anchor-end-nl`.

This fixes an incorrect `AST_ANALYSIS_ERROR_UNSUPPORTED` case in the
native dialect, where '$' was being treated like `\z` in PCRE, then
getting an unsupported error related to captures. This fixes the
test failure in tests/endids/endids10_minimise_partial_overlap.c.

silentbicycle committed:
Regenerate the parser for the native dialect.

silentbicycle committed:
ast_analysis: Fix ALT reverse anchoring case, likely merge error.

This looks like a merge error. The proper behavior here is to skip
unsupported subtrees and only fail as unsatisfiable if none are, but it
was returning immediately. This broke the `a|_$[^b]` test case in
`tests/capture/capture_test_case_list.c`.

silentbicycle committed:
Fix UB warning for signed int promotion with the negated flags.

silentbicycle committed:
Update test case: We can handle this correctly now.

Instead of rejecting this as unsupported, it should just prune the
unsatisfiable side of the `|` and produce the same FSM as `^a`.

silentbicycle committed:
ast_make_expr_alt: Ensure field is init'd to 0.

silentbicycle committed:
re_capvm_compile.c: Comment changes:

Rather than using numbered label names ("j2", etc.) use something more
descriptive.

Use "{x,x}" for repetition comment, to make bounds explicit.

silentbicycle committed:
Bugfix: Only check flag on union's alt if it's an AST_EXPR_ALT.
